server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java

```diff
@@ -407,12 +475,27 @@ public String executor() {
     }
 
     public void sendValidateJoinRequest(DiscoveryNode node, ClusterState state, ActionListener<TransportResponse.Empty> listener) {
-        transportService.sendRequest(
-            node,
-            VALIDATE_JOIN_ACTION_NAME,
-            new ValidateJoinRequest(state),
-            new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
-        );
+        try {
+            BytesReference bytes;
+            if (clusterStateRefreshInterval == 0) {
+                bytes = CompressionHelper.serializeClusterState(state, node, true);
+            } else {
+                bytes = serializedStates.computeIfAbsent(
+                    node.getVersion(),
+                    key -> CompressionHelper.serializeClusterState(state, node, true)
+                );
+            }
+            final BytesTransportRequest request = new BytesTransportRequest(bytes, node.getVersion());
+            transportService.sendRequest(
+                node,
+                VALIDATE_COMPRESSED_JOIN_ACTION_NAME,
+                request,
+                new ActionListenerResponseHandler<>(listener, i -> Empty.INSTANCE, ThreadPool.Names.GENERIC)
```

Comment:
ClusterManager should choose between the regular ValidateJoinRequest or compressed one based on the node version (you can get this by checking stream version). If it always sends compressed one then data nodes on older version wont be able to parse it.

Reply:
Agreed, thanks for catching this. I'll add a check for 3.0 version to send compressed state.

Reply:
Added a check for 2.8 considering other ongoing PRs.

Reply:
Not sure what you mean by "other ongoing PRs"? I think the right way to do this is to add an "on or after 3.0" check, then backport this and change it to "on or after 2.8" in the 2.x backport PR, then immediately open another PR on main to change this to be "on or after 2.8" once the backport is merged. It's a bit cumbersome but it does guarantee that all tests will pass in every PR. I believe the current PR revision will attempt to send the compressed version to 2.8, which won't know how to parse it.